internal/controller/clientprofile_controller.go

```diff
@@ -0,0 +1,337 @@
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"slices"
+	"sync"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	csiv1a1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	"github.com/ceph/ceph-csi-operator/internal/utils"
+)
+
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=clientprofile,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=clientprofile/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=clientprofile/finalizers,verbs=update
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=cephconnection,verbs=get;list;watch;update
+
+// ClientProfileReconciler reconciles a ClientProfile object
+type ClientProfileReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// A local reconcile object tied to a single reconcile iteration
+type ClientProfileReconcile struct {
+	ClientProfileReconciler
+
+	ctx           context.Context
+	log           logr.Logger
+	clientProfile csiv1a1.ClientProfile
+	cephConn      csiv1a1.CephConnection
+	cleanUp       bool
+}
+
+// csiClusterRrcordInfo represent the structure of a serialized csi record
+// in Ceph CSI's config, configmap
+type csiClusterInfoRecord struct {
+	ClusterId string   `json:"clusterID,omitempty"`
+	Monitors  []string `json:"monitors,omitempty"`
+	CephFs    struct {
+		SubvolumeGroup     string `json:"subvolumeGroup,omitempty"`
+		KernelMountOptions string `json:"kernelMountOptions"`
+		FuseMountOptions   string `json:"fuseMountOptions"`
+	} `json:"cephFS,omitempty"`
+	Rbd struct {
+		RadosNamespace string `json:"radosNamespace,omitempty"`
+		MirrorCount    int    `json:"mirrorCount,omitempty"`
+	} `json:"rbd,omitempty"`
+	Nfs          struct{} `json:"nfs,omitempty"`
+	ReadAffinity struct {
+		Enabled             bool     `json:"enabled,omitempty"`
+		CrushLocationLabels []string `json:"crushLocationLabels,omitempty"`
+	} `json:"readAffinity,omitempty"`
+}
+
+const (
+	cleanupFinalizer = "csi.ceph.com/cleanup"
+)
+
+var configMapUpdateLock = sync.Mutex{}
+
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=clientprofile,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=clientprofile/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=clientprofile/finalizers,verbs=update
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=cephconnection,verbs=get;list;watch;update
+//+kubebuilder:rbac:groups="",resources=configmap,verbs=get;list;watch;create;update;patch;delete
```

Comment:
resources should be plural. clientprofile to clientprofiles and cephconnection to cephconnections, configmap to configmaps

Reply:
Fixed now

Reply:
resolve.